libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,200 @@
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/analysis/FunctionCallGraph.h>
+#include <boost/range/adaptor/reversed.hpp>
+
+using namespace std;
+using namespace solidity::frontend;
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(Node const& _lhs, Node const& _rhs) const
+{
+	if (_lhs.index() != _rhs.index())
+		return _lhs.index() < _rhs.index();
+
+	if (std::holds_alternative<SpecialNode>(_lhs))
+		return std::get<SpecialNode>(_lhs) < std::get<SpecialNode>(_rhs);
+	return std::get<ASTNode const*>(_lhs)->id() < std::get<ASTNode const*>(_rhs)->id();
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(Node const& _lhs, int64_t _rhs) const
+{
+	solAssert(!std::holds_alternative<SpecialNode>(_lhs), "");
+
+	return std::get<ASTNode const*>(_lhs)->id() < _rhs;
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(int64_t _lhs, Node const& _rhs) const
+{
+	solAssert(!std::holds_alternative<SpecialNode>(_rhs), "");
+
+	return _lhs < std::get<ASTNode const*>(_rhs)->id();
+}
+
+unique_ptr<FunctionCallGraphBuilder::ContractCallGraph> FunctionCallGraphBuilder::create(ContractDefinition const& _contract)
+{
+	FunctionCallGraphBuilder builder(_contract);
+
+	// Create graph for constructor, state vars, etc
+	builder.m_currentNode = SpecialNode::EntryCreation;
+	builder.m_currentDispatch = SpecialNode::InternalCreationDispatch;
+
+	for (ContractDefinition const* contract: _contract.annotation().linearizedBaseContracts | boost::adaptors::reversed)
+	{
+		for (auto const* stateVar: contract->stateVariables())
+			stateVar->accept(builder);
+
+		for (auto arg: contract->baseContracts())
+			arg->accept(builder);
+
+		if (contract->constructor())
+		{
+			builder.add(*builder.m_currentNode, contract->constructor());
+			contract->constructor()->accept(builder);
+			builder.m_currentNode = contract->constructor();
+		}
+	}
+
+	builder.m_currentNode.reset();
+	builder.m_currentDispatch = SpecialNode::InternalDispatch;
+
+	// Create graph for all publicly reachable functions
+	for (auto& [hash, functionType]: _contract.interfaceFunctionList())
+	{
+		(void)hash;
+		if (auto const* funcDef = dynamic_cast<FunctionDefinition const*>(&functionType->declaration()))
+			if (!builder.m_graph->edges.count(funcDef))
+				builder.visitCallable(funcDef);
+
+		// Add all external functions to the RuntimeDispatch
+		builder.add(SpecialNode::Entry, &functionType->declaration());
```

Comment:
If you set ``m_currentNode = SpecialNode::Entry;`` before the loop you can skip doing this each time and have ``visitCallable`` take care of it, can't you? And that way you might even be able to assert that you have a previous node in ``visitCallable`` whenever it's a direct call?

Reply:
That is true, but there is the if condition before we call the function and it's checking only if we already have that function as key (it might have been added earlier from the constructor/state var init etc). We want to add an edge from `Entry` either way..

Reply:
Right, ok, that's a point...

Reply:
Probably also applies to my suggestion in the other thread.

Reply:
One other difference is that in the snippet above `functionType->declaration()` could be a `VariableDeclaration`. My tests were actually failing on that because I thought that we're supposed to ignore public getters but this was adding them anyway.

For now I have changed it to really skip public getters and I have also moved the other condition into `visitCallable()`. So I can actually apply this suggestion now.